File: bench/prime-field-matrix-comparison.m
// Matched Magma benchmark for Native Kernel v8 prime-field matrices.

sizes := [16, 32, 64, 128, 256];
fields := [
    <"u32", 65521>,
    <"u61", 2305843009213693951>
];

for field_data in fields do
    field_name := field_data[1];
    prime := field_data[2];
    field := GF(prime);
    for size in sizes do
        if size le 16 then
            repetitions := 10000;
        elif size le 32 then
            repetitions := 3000;
        elif size le 64 then
            repetitions := 500;
        elif size le 128 then
            repetitions := 100;
        else
            repetitions := 20;
        end if;
        source_values := [
            (field!(row + column - 1))^-1
            : row, column in [1..size]
        ];
        right_values := [
            field!(row * (column + 1))
            : row in [1..size], column in [1..4]
        ];

        sources := [
            Matrix(field, size, size, source_values)
            : repetition in [1..repetitions]
        ];
        started := Cputime();
        for repetition in [1..repetitions] do
            answer := Rank(sources[repetition]);
        end for;
        elapsed := 1000 * Cputime(started) / repetitions;
        print "RESULT", field_name, size, "rank", RealField(20)!elapsed;

        sources := [
            Matrix(field, size, size, source_values)
            : repetition in [1..repetitions]
        ];
        started := Cputime();
        for repetition in [1..repetitions] do
            answer := Determinant(sources[repetition]);
        end for;
        elapsed := 1000 * Cputime(started) / repetitions;
        print "RESULT", field_name, size, "determinant", RealField(20)!elapsed;

        sources := [
            Matrix(field, size, size, source_values)
            : repetition in [1..repetitions]
        ];
        started := Cputime();
        for repetition in [1..repetitions] do
            answer := EchelonForm(sources[repetition]);
        end for;
        elapsed := 1000 * Cputime(started) / repetitions;
        print "RESULT", field_name, size, "echelon", RealField(20)!elapsed;

        sources := [
            Matrix(field, size, size, source_values)
            : repetition in [1..repetitions]
        ];
        rights := [
            Matrix(field, size, 4, right_values)
            : repetition in [1..repetitions]
        ];
        started := Cputime();
        for repetition in [1..repetitions] do
            answer := Transpose(Solution(
                Transpose(sources[repetition]),
                Transpose(rights[repetition])
            ));
        end for;
        elapsed := 1000 * Cputime(started) / repetitions;
        print "RESULT", field_name, size, "solve-4", RealField(20)!elapsed;
    end for;
end for;

quit;

File: bench/prime-field-matrix-comparison.jl
#!/usr/bin/env julia

using Nemo
using Printf
using Statistics

const SIZES = [16, 32, 64, 128, 256]
const PRIMES = [("u32", 65521), ("u61", 2305843009213693951)]

function repetitions(size)
    size <= 16 && return 50
    size <= 32 && return 20
    size <= 64 && return 8
    size <= 128 && return 3
    return 1
end

function measure(operation, count)
    for _ in 1:3
        operation()
    end
    samples = Float64[]
    for _ in 1:7
        elapsed = @elapsed for _ in 1:count
            operation()
        end
        push!(samples, 1000 * elapsed / count)
    end
    return median(samples)
end

function benchmark(field_name, prime, size)
    field = GF(prime)
    source = matrix(
        field,
        size,
        size,
        [inv(field(row + column - 1))
         for row in 1:size for column in 1:size],
    )
    right = matrix(
        field,
        size,
        4,
        [field(row * (column + 1))
         for row in 1:size for column in 1:4],
    )
    count = repetitions(size)
    operations = [
        ("rank", () -> rank(source)),
        ("determinant", () -> det(source)),
        ("echelon", () -> rref(source)),
        ("solve-4", () -> solve(source, right; side = :right)),
    ]
    for (name, operation) in operations
        milliseconds = measure(operation, count)
        @printf("RESULT %s %d %s %.12f\n",
            field_name, size, name, milliseconds)
    end
end

for (field_name, prime) in PRIMES
    for size in SIZES
        benchmark(field_name, prime, size)
    end
end
